iceoryx_dds/source/iceoryx_dds/dds/fastdds_data_writer.cpp

```diff
@@ -0,0 +1,165 @@
+    }
+    if (m_participant != nullptr)
+    {
+        eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->delete_participant(m_participant);
+    }
+    LogDebug() << m_logName << "Destroyed FastDDSDataWriter";
+}
+
+void iox::dds::FastDDSDataWriter::connect() noexcept
+{
+    if (!m_isConnected.load(std::memory_order_relaxed))
+    {
+        // Create the DomainParticipant
+        eprosima::fastdds::dds::DomainParticipantQos participantQos;
+        m_participant =
+            eprosima::fastdds::dds::DomainParticipantFactory::get_instance()->create_participant(0, participantQos);
+
+        // Register the type
+        m_type.register_type(m_participant);
+
+        // Create topic
+        auto m_topicname =
+            "/" + std::string(m_serviceId) + "/" + std::string(m_instanceId) + "/" + std::string(m_eventId);
+        m_topic = m_participant->create_topic(m_topicname, "Mempool::Chunk", eprosima::fastdds::dds::TOPIC_QOS_DEFAULT);
+
+        // Create publisher
+        m_publisher = m_participant->create_publisher(eprosima::fastdds::dds::PUBLISHER_QOS_DEFAULT, nullptr);
+
+        // Create datawriter
+        eprosima::fastdds::dds::DataWriterQos dataWriterQos = eprosima::fastdds::dds::DATAWRITER_QOS_DEFAULT;
+        dataWriterQos.endpoint().history_memory_policy =
+            eprosima::fastrtps::rtps::PREALLOCATED_WITH_REALLOC_MEMORY_MODE;
+        m_dataWriter = m_publisher->create_datawriter(m_topic, dataWriterQos, &m_listener);
+
+        // Save state
+        m_isConnected.store(true, std::memory_order_relaxed);
+        LogDebug() << m_logName << "Connected to topic: " << m_topicname;
+    }
+}
+
+void iox::dds::FastDDSDataWriter::write(
+        const uint8_t* const bytes,
+        const uint64_t size) noexcept
+{
+    try
+    {
+        Mempool::Chunk data{static_cast<uint32_t>(size), const_cast<uint8_t*>(bytes)};
+        if (!m_dataWriter->write(&data))
+        {
+            LogError() << m_logName << "Cannot send data";
+        }
+    }
+    catch(const std::exception& e)
+    {
+        LogError() << m_logName << "write exception: " << e.what();
+    }
+}
+
+iox::capro::IdString_t iox::dds::FastDDSDataWriter::getServiceId() const noexcept
+{
+    return m_serviceId;
+}
+
+iox::capro::IdString_t iox::dds::FastDDSDataWriter::getInstanceId() const noexcept
+{
+    return m_instanceId;
+}
+
+iox::capro::IdString_t iox::dds::FastDDSDataWriter::getEventId() const noexcept
+{
+    return m_eventId;
+}
+
+void iox::dds::FastDDSDataWriter::waitForReaderDiscovery(
+        uint16_t readersCount)
+{
+    std::unique_lock<std::mutex> lk(m_mutex);
+    m_readerDiscoveryCv.wait(
+        lk,
+        [&]
+        {
+            return (m_readerDiscoveryCount >= readersCount);
+        });
+
+    lk.unlock();
+    m_readerDiscoveryCv.notify_one();
+}
+
+iox::dds::FastDDSDataWriter::FastDDSDataWriterListener::FastDDSDataWriterListener(
+        FastDDSDataWriter* dataWriter)
```

Comment:
What happens when `dataWriter == nullptr` maybe we should here also use `cxx::not_null`.